keps/976-plain-pods/README.md

```diff
@@ -0,0 +1,678 @@
+as elastic jobs.
+
+A recent enhancement to Kubernetes Pods, scheduling gates, introduced in 1.26 as Alpha, and 1.27 as
+Beta, allows an external controller to prevent kube-scheduler from scheduling Pods. Kueue can make
+use of this API to implement queuing semantics for Pods.
+
+### Goals
+
+- Support queueing of individual Pods.
+- Support queueing of groups of Pods of fixed size, identified by a common label or annotation.
+- Opt-in or opt-out Pods from specific namespaces from queuing.
+
+### Non-Goals
+
+- Support for [dynamic reclaiming quota](https://github.com/kubernetes-sigs/kueue/issues/78)
+
+  This feature is incompatible with supporting Pod replacements without knowing the behavior of a
+  parent controller for the Pods.
+
+- Support for partial-admission.
+
+  Since all pods are already created, an implementation of partial admission would imply the
+  deletion of some pods. It is not clear if this matches users expectations, as opposed to support
+  for elastic groups.
+
+- Support elastic groups of Pods, where the number of Pods changes after the job started.
+
+  While these jobs are one of the motivations for this KEP, the current proposal doesn't support
+  them. These jobs can be addressed in follow up KEPs.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+#### Story 1
+
+As an platform developer, I can queue plain Pods. I just add a queue name to the Pods through a
+label.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: foo
+  namespace: pod-namespace
+  labels:
+    kueue.x-k8s.io/queue-name: user-queue
+spec:
+  containers:
+  - name: job
+    image: hello-world
+    resources:
+      requests:
+        cpu: 1m
+```
+
+#### Story 2
+
+As a platform developer, I can queue groups of Pods that share the same shape (Pod specs).
+In addition to the queue name, I can specify how many Pods belong to the group.
+
+<<[UNRESOLVED usability of labels and annotations ]>>
+Could the split among labels and annotations cause user mistakes?
+Note that integers cannot be used as label values.
+<<[/UNRESOLVED]>>
+
+The pods of a job with a single spec (similar to Pods in an Indexed Job), look like follows:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod-index-0
+  namespace: pod-namespace
+  labels:
+    kueue.x-k8s.io/queue-name: user-queue
+    kueue.x-k8s.io/pod-group-name: pod-group
+  annotations:
+    kueue.x-k8s.io/pod-group-count: "10"
+spec:
+  containers:
+  - name: job
+    image: hello-world
+    resources:
+      requests:
+        cpu: 1m
+```
+
+#### Story 3
```

Comment:
So reading this I do wonder why would we not suggest something like JobSet?

There are a lot of downsides to using naked Pods and I'm not sure you call it out that this should be only used in the case that you can't migrate to a CRD. 

I think items like self-healing, retries, state handling are really difficult problems to solve with a pod only API.  I understand that you feel the same way about this but I'm sure there are reasons to not migrate.  

Maybe this is sorta an implementation detail for documentaiton but I do worry that this should be more of bandaid solution rather than a permanent solution.

Reply:
I fully agree with you! In an ideal world, users would be using Jobs and JobSets. But the motivation for this proposal is to meet users where they are today to ease their transition to Kueue.

Reply:
In the real world, migrating existing systems using naked pods to the system using JobSet or Job is hard since there are cases of the system depending on other systems or other systems depending on the system.

Reply:
Yea.  I say this as someone who works at an organization that uses naked pods.   There are a lot of things we are missing out on and we are getting requests for things like automatic retries, failure/success policies, and gang scheduling.  I have been trying to convince my org to adopt the Job API but I have been unsuccessful mostly for legacy reasons.  I just worry that this will continue to make the sell hard if we have an easy way to add queueing/essentially a PodSet API.  

I just worry that if we aren't strong enough in wording that we would just end up prolonging the transition to Jobs/Jobsets since we allow for the use of pods.  I guess if people are using a controller to manage those pods that isn't a bad thing.  But supporting pods without any owner is a code smell and I just wonder if we should maybe mention that in this KEP?

Anyway I know you feel the same way so I want press too much.

Reply:
That being said, things like preemption and automatic retires wouldn't be here with naked pods so maybe we hope that would be sufficient to get people to migrate eventually.

Reply:
I feel your frustration :(
We are already limiting the usefulness of the naked pods integration, for example, without partial admission or dynamic reclaiming. And keeping the pod finalizers until all the pods finish. I hope this is enough to keep this integration just as a stepping stone. We certainly would never support retries in Kueue. There is Job for that :)

Reply:
Added to Non-goals

Reply:
Sounds good.

Reply:
+1 that folks should adopt Jobs but sadly large ML platforms have been built that rely on naked pods. Those ML platforms have very small appetite for change and risk and would rather built a custom controller that does part of what Kueue does than to adopt Kueue with Job.

Thank you for willing to meet users where they are even though it doesn't make sense from a technology pov.

Reply:
I think to be fair there are mature external controllers that were created well before Kueue and to address gaps in the Jobs API. Supporting the transition of users of these controllers (which Kueue may not yet even fully cover) is an important investment and I'm glad to see this proposal.